fs/fs_zenfs.cc

```diff
@@ -620,12 +621,33 @@ IOStatus ZenFS::ReopenWritableFile(const std::string& fname,
   return OpenWritableFile(fname, file_opts, result, dbg, true);
 }
 
+/* Must hold files_mtx_ */
+void ZenFS::GetZenFSChildrenNoLock(const std::string& dir,
+                                   bool include_grandchildren,
+                                   std::vector<std::string>* result) {
+  for (auto it = files_.begin(); it != files_.end(); it++) {
+    std::filesystem::path file_path(it->first);
+    if (file_path.has_filename()) {
+      // Making sure that we end the dir with a seperator
+      std::string dir_with_terminating_seperator =
+          (std::filesystem::path(dir) / "").lexically_normal().string();
+      std::string file_dir =
+          (file_path.parent_path() / "").lexically_normal().string();
+      if (file_dir.rfind(dir_with_terminating_seperator, 0) == 0) {
+        if (include_grandchildren ||
+            file_dir.length() == dir_with_terminating_seperator.length()) {
+          result->push_back(file_path.filename().string());
+        }
+      }
+    }
+  }
+}
```

Comment:
```suggestion
void ZenFS::GetZenFSChildrenNoLock(const std::string& dir,
                                   bool include_grandchildren,
                                   std::vector<std::string> &result) {
  auto path_as_string_with_separator_at_end =
      [](std::filesystem::path const& path) {
        auto with_sep = path / "";
        return with_sep.lexically_normal().string();
      };

  auto string_starts_with = [](std::string const& string,
                               std::string const& needle) {
    return string.rfind(needle, 0) == 0;
  };

  std::string dir_with_terminating_seperator =
      path_as_string_with_separator_at_end(std::filesystem::path(dir));

  for (auto const& it : files_) {
    std::filesystem::path file_path(it.first);
    assert(file_path.has_filename());
    // if (!file_path.has_filename()) continue; // <- This should be impossible?

    std::string file_dir =
        path_as_string_with_separator_at_end(file_path.parent_path());

    if (string_starts_with(file_dir, dir_with_terminating_seperator)) {
      if (include_grandchildren ||
          file_dir.length() == dir_with_terminating_seperator.length()) {
        result.push_back(file_path.filename().string());
      }
    }
  }
}
```

Do we need to add the separator at end before comparison? Why can we not compare normalized paths without the terminating separator?

Reply:
I like the refactoring suggestions!

We need the separator at the end before comparison, otherwise we would find false children if the dir string is a filename (not a directory) that is a substring of an existing directory.
Normalising a path without an explicit separator at the end of the path does not add the separator at the end of a directory https://en.cppreference.com/w/cpp/filesystem/path/lexically_normal.

Reply:
We could check that `dir` is a directory in `target()` first, then drop the separator append?

Reply:
```c++
void ZenFS::GetZenFSChildrenNoLock(const std::string& dir,
                                   bool include_grandchildren,
                                   std::vector<std::string>& result) {

  bool target_is_dir;
  IOOptions options;
  IODebugContext context;
  IOStatus status = target()->IsDirectory(dir, options, &target_is_dir, &context);
  // TODO if (!status.ok()) return status;
  if (!target_is_dir) {
    // TODO Return error
  }

  auto string_starts_with = [](std::string const& string,
                               std::string const& needle) {
    return string.rfind(needle, 0) == 0;
  };

  for (auto const& it : files_) {
    std::filesystem::path file_path(it.first);
    assert(file_path.has_filename());
    // if (!file_path.has_filename()) continue; // <- This should be impossible?

    std::string file_dir =
      file_path.parent_path();

    if (string_starts_with(file_dir, dir)) {
      if (include_grandchildren ||
          file_dir.length() == dir.length()) {
        result.push_back(file_path.filename().string());
      }
    }
  }
}
```

Reply:
This would not cover the following case:
Let's assume we call `GetZenFSChildrenNoLock("/path/exampledir", true, result_vector)` where `/path/exampledir` is a directory in the aux directory and `files_` contains a ZenFS file called `/path/exampledir2/examplefile`. 
The check if `/path/exampledir` is a directory would pass and `if (string_starts_with("/path/exampledir2", dir))` would also pass which means we would falsely consider `/path/exampledir2/examplefile` as a child of `/path/exampledir/`

Reply:
But then we could check if character after the prefix we search for is a separator. But maybe that is not better than what we have.

Reply:
Yep I think this is essentially the same check.